### ISSUE TITLE
force upgrade to specific version

### ISSUE THREAD

**Issue by aDotNetDeveloper** [aaronpowell/Chauffeur#51]
Hi,

just a general 'how can we' question on umbraco upgrade via a delivery;

How could a delivery access this
`        public virtual bool Execute(Database database, bool isUpgrade = true)
        {
            return Execute(database, database.GetDatabaseProvider(), isUpgrade);
        }
`

which is part of `MigrationRunner` and in inside the DatabaseContext internal method `UpgradeSchemaAndData` ?

**Comment by aaronpowell** [comment]
Upgrading is a really difficult one to tackle in Chauffeur. The problem that you hit is that you can't really have NuGet upgrade as a step, so if you say, had a `delivery` that did `install y`, a few other deliveries after it that add doc types, etc. then you want to do an Umbraco upgrade you would upgrade with NuGet.

Now if you deleted everything locally and re-cloned the repo you're `install y` step wouldn't know to install Umbraco 7.5.0, it'd use whatever is in the NuGet package (say, Umbraco 7.6.0).

Basically I don't know how to solve this.

Hopefully that gives you some insight, or I might've completely missed the point what you were asking 😛.

On the topic of accessing internal methods, well you're going to be going down the rabbit hole of reflection!

**Comment by aaronpowell** [comment]
Yep ok, so it is somewhat like the problem I thought.

The issue you hit is what happens when you clone onto a new machine, how does the `UpgradeDeliverable` react?

Thinking it through a little bit it could work like follows:

- Require a _target version_ or default to what's in the `web.config`
- `UpgradeDeliverable` will then look into the `UmbracoMigration` table to find the current DB version
- If the version of the DB matches the target then just exit
- If the version of the DB doesn't match the target run the Umbraco migration API to the target

Not quite sure how I'll do any automation tests for that though 😛

**Comment by aDotNetDeveloper** [comment]
yep, exactly the process I had outlined in my custom deliverable... although I was being more brutal and have a delivery command of;

    upgradedb 7.5 7.6

until part 4, performing the upgrade

any ideas the best way to implement something like this? reflection or not.

**Comment by aaronpowell** [comment]
I try and avoid reflection where possible, primarily because it is quite brittle (and in this case it would be **very** hard to write an automation test to cover it so the change of it breaking without being caught is high).

I'd want to explore more into the migrations API in Umbraco, see if there's not something public that can be used as an entry point, even if it's not intended to be used, which, after all, is how most of this works 😛.

**Comment by aDotNetDeveloper** [comment]
ok, great keen to try something

**Comment by aaronpowell** [comment]
So a quick poke around in the Umbraco source code I think what we'd need to do is use the `MigrationRunner` class (which is `public`).

Here's the constructor:

```
public MigrationRunner(IMigrationEntryService migrationEntryService, ILogger logger, SemVersion currentVersion, SemVersion targetVersion, string productName, params IMigration[] migrations)
```

Then you use the `Execute` method that is on it (the `isUpgrade` parameter is whether you want Umbraco do treat it as an upgrade or downgrade operation).

**Comment by aDotNetDeveloper** [comment]
Migrations param had me stumped.

**Comment by aaronpowell** [comment]
If you look into the Umbraco source and the usage of that `ctor` what you'll find is that it's never using the `params` argument.

**Comment by stevetemple** [comment]
I was having some fun with upgrades too. I've knocked this up, but it'll just upgrade to the current version by running all the migrations from whatever the current version is https://github.com/aaronpowell/Chauffeur/pull/85

The only issue I have is getting any meaningful unit testing of it as its pretty much all in Umbraco so coverage has dropped

**Comment by aaronpowell** [comment]
Awesome @stevetemple! I've added some comments on the PR for you, let's work together to get this going 🤩

**Comment by aaronpowell** [comment]
And release 1.4.0 has been cut with this in it! Go community! 😁
